### ISSUE TITLE
map.queryRenderedFeatures vs. onHover event.features

### ISSUE THREAD

**Issue by amber-chen** [visgl/react-map-gl#1337]
For our `onHover` function, we have multiple different features (that may or not be there depending on certain controls and settings for the map) that we want to handle on the map.

From my understanding, we can use `interactiveLayerIds` to set what features are detected on any of the hover/click events, and then use `event.features` to check if it matches the case for layer A if so then handle it, or else check for the other interested layers in `interactiveLayerIds`.

However, this way, the `event.features` is empty a lot of the times. Is this expected? 

As a workaround, we use `map.queryRenderedFeatures` to query for each of the cases in which we want to handle hover functionality for layer A, B, and so on. This way seems to consistently work, but I was wondering if `event.features` should return all the features under the pointer's position, if so - is there any reason why it might not return everything? I was thinking it could be related to `interactiveLayerIds`?

This thread here - https://github.com/visgl/react-map-gl/issues/1184 leads me to believe I should just be using `map.queryRenderedFeatures` to handle each of the cases, but I'm wondering if that is costly to do it every time, and why I'm not getting just the features I care about from the event itself?

Any clarifications are helpful!

**Comment by Pessimistress** [comment]
`event.features` is populated by calling `queryRenderedFeatures`. If it doesn't work as expected, please create a CodeSandbox that reproduces the issue.

**Comment by amber-chen** [comment]
[CodeSandbox](https://codesandbox.io/s/focused-blackburn-71bxp?file=/src/App.tsx) I was thinking it could be because of the ref and/or the hooks that I need to save the instance of the map. event.features is null/empty when hovering over the 2 red pins.

**Comment by Pessimistress** [comment]
Hmm looks like the map component no longer works with functional refs. If you use:

```js
const mapRef = useRef();
return <MapGL ref={mapRef} ... />;
```

`event.features` are populated correctly.

I will look into this as this was an unintended breaking change. However, React has explicitly marked functional refs to be deprecated, and you should use ref hooks the way it is documented.

**Comment by amber-chen** [comment]
It looks like declaring `const mapRef = useRef()` inside the Map container component works, but when declaring it in the context, and using the context hook in the Map container component, the `mapRef` is null. Is that expected?

Even when I declare the below, the features still aren't identified as expected.

`const mapRef = useRef();
return <MapGL ref={mapRef} ... />;`

I modified the sandbox code to show an example: https://codesandbox.io/s/focused-blackburn-71bxp?file=/src/App.tsx

**Comment by Pessimistress** [comment]
Your new sandbox shows `event.features` returning the same result as `queryRenderedFeatures`.

**Comment by amber-chen** [comment]
Oh duh I see it working now, but it only works when I declare the ref within the same Map Container component and not when i declare the ref in the context hook (to use elsewhere outside the direct Map component), is that expected?

**Comment by Pessimistress** [comment]
Your context hook has a default of `mapRef: null`.

**Comment by Pessimistress** [comment]
A fix for function refs is landed in 5.3.5/6.1.5
